codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java

```diff
@@ -0,0 +1,252 @@
+            }
+
+            @Override
+            public void onStreamClosed(Http2Stream stream) {
+                state(stream).close();
+            }
+        });
+    }
+
+    @Override
+    public void updateStreamableBytes(StreamState streamState) {
+        State state = state(streamState.stream());
+        state.updateStreamableBytes(streamState.streamableBytes(), streamState.hasFrame());
+    }
+
+    @Override
+    public boolean distribute(int maxBytes, Writer writer) throws Http2Exception {
+        checkNotNull(writer, "writer");
+
+        // Allocate the bytes to the streams.
+        long allocated = allocateBytes(maxBytes);
+
+        long unallocatedStreamableBytes = totalStreamableBytes - allocated;
+
+        // Need to write even if maxBytes == 0 in order to handle the case of empty frames.
+        writeVisitor.writeAllocatedBytes(writer);
+
+        return unallocatedStreamableBytes > 0;
+    }
+
+    /**
+     * Allocates the given bytes across the streams.
+     *
+     * @param maxBytes The maximum number of bytes to allocate.
+     * @return the total number of bytes actually allocated.
+     */
+    private long allocateBytes(int maxBytes) {
+        if (queue.isEmpty() || maxBytes <= 0) {
+            return 0L;
+        }
+
+        long allocated = 0;
+        long unallocated = totalStreamableBytes;
+        int allocationChunk = max(minAllocationChunk, (int)(unallocated / queue.size()));
+        while (maxBytes > 0 && !queue.isEmpty()) {
+            //int allocationChunk = max(minAllocationChunk, (int)(unallocated / queue.size()));
+
+            // Remove the head of the queue.
+            State state = queue.remove();
+            state.enqueued = false;
+
+            // Allocate as much data as we can.
+            int chunk = min(allocationChunk, min(maxBytes, state.unallocated()));
+            state.allocate(chunk);
+            maxBytes -= chunk;
+            allocated += chunk;
+            unallocated -= chunk;
+
+            if (state.unallocated() > 0) {
+                // The stream still has unallocated bytes, add it back to the end of the queue.
+                state.enqueue();
+            }
+        }
+
+        return allocated;
+    }
+
+    private State state(Http2Stream stream) {
+        return checkNotNull(stream, "stream").getProperty(stateKey);
+    }
+
+    private class EnqueueStrategy {
+        void enqueue(State state) {
+            queue.addLast(state);
+            state.enqueued = true;
+        }
+    }
+
+    private final EnqueueStrategy defaultEnqueueStrategy = new EnqueueStrategy();
+
+    /**
+     * The first time a state is enqueued, it will be added to the head of the queue. After
+     * that it will be added to the tail.
+     */
+    private final EnqueueStrategy firstEnqueueStrategy = new EnqueueStrategy() {
+        @Override
+        void enqueue(State state) {
+            queue.addFirst(state);
+            state.enqueued = true;
```

Comment:
They're different. The first time it's enqueued, it is added to the head of the queue ... after that it's added to the end.